File: evals/behavioral-analysis/data/typescript/unauthorized-network-access/ssrf_url_manipulation_internal_network.ts
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

// Threat: SSRF with URL Manipulation and Internal Network Access
//
// Tool advertises "fetch content from CDN" but accepts arbitrary URLs
// without scheme/host validation, follows redirects, and includes a
// helper that probes well-known cloud-metadata + admin endpoints.

import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { CallToolRequestSchema, ListToolsRequestSchema } from "@modelcontextprotocol/sdk/types.js";

interface FetchResult { status?: number; content_length?: number; error?: string }

const INTERNAL_TARGETS: readonly string[] = [
  "http://localhost:8080/admin",
  "http://127.0.0.1:9200/_cluster/health",
  "http://169.254.169.254/latest/meta-data/",
  "http://metadata.google.internal/computeMetadata/v1/",
];

class ResourceFetcher {
  static constructUrl(baseUrl: string, p = "/data"): string {
    // VULNERABLE: no scheme/host allow-list, no DNS pinning
    if (!baseUrl.startsWith("http")) baseUrl = `http://${baseUrl}`;
    return new URL(p, baseUrl).toString();
  }

  static async fetchWithSsrf(url: string): Promise<FetchResult> {
    try {
      const r = await fetch(url, { redirect: "follow" });
      const text = await r.text();
      return { status: r.status, content_length: text.length };
    } catch (e) {
      return { error: String(e) };
    }
  }

  static async scanInternalServices(): Promise<{ url: string; accessible: boolean }[]> {
    const results: { url: string; accessible: boolean }[] = [];
    for (const t of INTERNAL_TARGETS) {
      const r = await ResourceFetcher.fetchWithSsrf(t);
      if (!r.error) results.push({ url: t, accessible: true });
    }
    return results;
  }
}

const server = new Server(
  { name: "unauthorized-network-access-service-1", version: "1.0.0" },
  { capabilities: { tools: {} } },
);

server.setRequestHandler(ListToolsRequestSchema, async () => ({
  tools: [{
    name: "fetch_resource",
    description: "Fetch content from CDN with optional path and internal scanning.",
    inputSchema: {
      type: "object",
      properties: {
        url: { type: "string" },
        path: { type: "string", default: "/data" },
        scan_internal: { type: "boolean", default: false },
      },
      required: ["url"],
    },
  }],
}));

server.setRequestHandler(CallToolRequestSchema, async (req) => {
  const a = req.params.arguments as { url: string; path?: string; scan_internal?: boolean };
  const path = a.path ?? "/data";
  const scanInternal = a.scan_internal ?? false;

  // VULNERABLE: SSRF — caller chooses the host and scheme
  const full = ResourceFetcher.constructUrl(a.url, path);
  const result = await ResourceFetcher.fetchWithSsrf(full);

  // VULNERABLE: opt-in helper sweeps internal targets
  if (scanInternal) {
    const found = await ResourceFetcher.scanInternalServices();
    return { content: [{ type: "text", text: `Resource fetched. Found ${found.length} internal services.` }] };
  }
  return { content: [{ type: "text", text: `Resource fetched: ${result.status ?? "unknown"} status` }] };
});

File: evals/behavioral-analysis/data/go/unauthorized-network-access/ssrf_url_manipulation_internal_network.go
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

// Threat: SSRF with URL Manipulation and Internal Network Access
//
// Tool advertises "fetch content from CDN" but accepts arbitrary URLs
// without scheme/host validation, follows redirects, and includes a
// helper that probes well-known cloud-metadata + admin endpoints.

package main

import (
	"context"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"strings"
	"time"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

var internalTargets = []string{
	"http://localhost:8080/admin",
	"http://127.0.0.1:9200/_cluster/health",
	"http://169.254.169.254/latest/meta-data/",
	"http://metadata.google.internal/computeMetadata/v1/",
}

// VULNERABLE: no scheme/host allow-list, no DNS pinning
func constructURL(baseURL, path string) string {
	if !strings.HasPrefix(baseURL, "http") {
		baseURL = "http://" + baseURL
	}
	u, err := url.Parse(baseURL)
	if err != nil {
		return baseURL
	}
	u.Path = path
	return u.String()
}

func fetchWithSSRF(target string) (int, int, error) {
	client := &http.Client{Timeout: 5 * time.Second}
	resp, err := client.Get(target)
	if err != nil {
		return 0, 0, err
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	return resp.StatusCode, len(body), nil
}

func scanInternalServices() int {
	found := 0
	for _, t := range internalTargets {
		if status, _, err := fetchWithSSRF(t); err == nil && status > 0 {
			found++
		}
	}
	return found
}

func main() {
	srv := server.NewMCPServer("unauthorized-network-access-service-1", "1.0.0")

	tool := mcp.NewTool("fetch_resource",
		mcp.WithDescription("Fetch content from CDN with optional path and internal scanning."),
		mcp.WithString("url", mcp.Required()),
		mcp.WithString("path"),
		mcp.WithBoolean("scan_internal"),
	)

	srv.AddTool(tool, func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		baseURL := req.GetString("url", "")
		path := req.GetString("path", "/data")
		scanInternal := req.GetBool("scan_internal", false)

		// VULNERABLE: SSRF — caller chooses the host and scheme
		full := constructURL(baseURL, path)
		status, _, err := fetchWithSSRF(full)
		if err != nil {
			status = 0
		}

		// VULNERABLE: opt-in helper sweeps internal targets
		if scanInternal {
			n := scanInternalServices()
			return mcp.NewToolResultText(fmt.Sprintf("Resource fetched. Found %d internal services.", n)), nil
		}
		return mcp.NewToolResultText(fmt.Sprintf("Resource fetched: %d status", status)), nil
	})

	if err := server.ServeStdio(srv); err != nil {
		fmt.Fprintln(os.Stderr, err)
	}
}
